### ISSUE TITLE
issue with splitting volume

### ISSUE THREAD

**Issue by TuotuoLi** [facebookresearch/NSVF#9]
Amazing work, Thanks for sharing the source!

I run into some problem while training, after the first volume splitting, part of the volume that should be occupied become empty.
I am curious if this is a bug or because package version issue. I am using python 3.8, pytorch 1.60.

Thanks!
![IMG_0508](https://user-images.githubusercontent.com/5238154/96048589-a9cb7d00-0e2b-11eb-873e-f6109a68be1f.jpg)

**Comment by nitthilan** [comment]
I too have seen it after 5 epochs or after 25000 iterations. Am not sure why we see these patches.

**Comment by MultiPath** [comment]
Hi, can you provide the training script you used to reproduce this results?

**Comment by TuotuoLi** [comment]
python train.py ${DATASET} \
    --user-dir fairnr \
    --task single_object_rendering \
    --train-views "0..100" --view-resolution "800x800" \
    --max-sentences 1 --view-per-batch 2 --pixel-per-view 2048 \
    --no-preload \
    --sampling-on-mask 1.0 --no-sampling-at-reader \
    --valid-views "100..110" --valid-view-resolution "400x400" \
    --valid-view-per-batch 1 \
    --transparent-background "1.0,1.0,1.0" --background-stop-gradient \
    --arch nsvf_base \
    --initial-boundingbox ${DATASET}/bbox.txt \
    --use-octree \
    --raymarching-stepsize-ratio 0.125 \
    --discrete-regularization \
    --color-weight 128.0 --alpha-weight 1.0 \
    --optimizer "adam" --adam-betas "(0.9, 0.999)" \
    --lr 0.001 --lr-scheduler "polynomial_decay" --total-num-update 150000 \
    --criterion "srn_loss" --clip-norm 0.0 \
    --num-workers 0 \
    --seed 2 \
    --save-interval-updates 500 --max-update 150000 \
    --virtual-epoch-steps 5000 --save-interval 1 \
    --half-voxel-size-at  "5000,25000,75000" \
    --reduce-step-size-at "5000,25000,75000" \
    --pruning-every-steps 2500 \
    --keep-interval-updates 5 --keep-last-epochs 5 \
    --log-format simple --log-interval 1 \
    --save-dir ${SAVE} \
    --tensorboard-logdir ${SAVE}/tensorboard \
    | tee -a $SAVE/train.log


Here is the training script that I used to reproduced this. it happens around 5000 iter after splitting volume.
![volume](https://user-images.githubusercontent.com/5238154/96062609-f5415380-0e4a-11eb-9df9-6763f45c46c9.png)

**Comment by MultiPath** [comment]
> python train.py ${DATASET}
> --user-dir fairnr
> --task single_object_rendering
> --train-views "0..100" --view-resolution "800x800"
> --max-sentences 1 --view-per-batch 2 --pixel-per-view 2048
> --no-preload
> --sampling-on-mask 1.0 --no-sampling-at-reader
> --valid-views "100..110" --valid-view-resolution "400x400"
> --valid-view-per-batch 1
> --transparent-background "1.0,1.0,1.0" --background-stop-gradient
> --arch nsvf_base
> --initial-boundingbox ${DATASET}/bbox.txt
> --use-octree
> --raymarching-stepsize-ratio 0.125
> --discrete-regularization
> --color-weight 128.0 --alpha-weight 1.0
> --optimizer "adam" --adam-betas "(0.9, 0.999)"
> --lr 0.001 --lr-scheduler "polynomial_decay" --total-num-update 150000
> --criterion "srn_loss" --clip-norm 0.0
> --num-workers 0
> --seed 2
> --save-interval-updates 500 --max-update 150000
> --virtual-epoch-steps 5000 --save-interval 1
> --half-voxel-size-at "5000,25000,75000"
> --reduce-step-size-at "5000,25000,75000"
> --pruning-every-steps 2500
> --keep-interval-updates 5 --keep-last-epochs 5
> --log-format simple --log-interval 1
> --save-dir ${SAVE}
> --tensorboard-logdir ${SAVE}/tensorboard
> | tee -a $SAVE/train.log
> 
> Here is the training script that I used to reproduced this. it happens around 5000 iter after splitting volume.
> ![volume](https://user-images.githubusercontent.com/5238154/96062609-f5415380-0e4a-11eb-9df9-6763f45c46c9.png)

Thanks, will re-run on my side and let you tonight.

**Comment by MultiPath** [comment]
Also, it would be better if you have the training logs. Especially logs when splitting voxel happens. Thanks

**Comment by TuotuoLi** [comment]
Attached please find the log.
[train.LOG](https://github.com/facebookresearch/NSVF/files/5381960/train.LOG)
Lines that might be of interet:
`2020-10-14 18:12:47 | INFO | valid | epoch 001 | valid on 'valid' subset | loss 2 | vox 0 | stp 0 | tvo 0 | asf 7 | ash 26 | nvo 155 | color 0 | alpha 0 | ssim 0 | psnr 27 | wps 1.6 | wpb 1 | bsz 1 | num_updates 5000 | best_loss 2
2020-10-14 18:12:47 | INFO | fairseq.checkpoint_utils | saved checkpoint /home/tuotuoli/work/output/NSVF/Bike/checkpoint1.pt (epoch 1 @ 5000 updates, score 2.0) (writing took 0.06901660500079743 seconds)
2020-10-14 18:12:47 | INFO | fairnr.modules.encoder | pruning...
2020-10-14 18:12:47 | INFO | fairnr.modules.encoder | pruning done. # of voxels before: 155, after: 148 voxels
2020-10-14 18:12:47 | INFO | fairnr.models.nsvf | half the global voxel size 0.1600 -> 0.0800
2020-10-14 18:12:47 | INFO | fairnr.modules.encoder | splitting...
Building EasyOctree done. total #nodes = 367, terminal #nodes = 148 (time taken 0.041932 s)
2020-10-14 18:12:47 | INFO | fairnr.modules.encoder | splitting done. # of voxels before: 148, after: 1184 voxels
2020-10-14 18:12:47 | INFO | fairnr_cli.train | reset the trainer at 5000
2020-10-14 18:12:47 | INFO | fairseq.trainer | NOTE: your device may support faster training with --fp16
2020-10-14 18:12:47 | INFO | fairnr.models.nsvf | reduce the raymarching step size 0.0200 -> 0.0100
Building EasyOctree done. total #nodes = 2479, terminal #nodes = 1184 (time taken 0.345659 s)
2020-10-14 18:12:48 | INFO | train_inner | epoch 002:      1 / 5000 loss=37, vox=0, stp=0, tvo=0, asf=42, ash=42, nvo=1184, color=0, alpha=0, wps=0.1, ups=0.07, wpb=2, bsz=1, num_updates=5001, lr=0.00096666, gnorm=106, train_wall=1, wall=1848`

**Comment by MultiPath** [comment]
> Attached please find the log.
> [train.LOG](https://github.com/facebookresearch/NSVF/files/5381960/train.LOG)
> Lines that might be of interet:
> 2020-10-14 18:12:47 | INFO | valid | epoch 001 | valid on 'valid' subset | loss 2 | vox 0 | stp 0 | tvo 0 | asf 7 | ash 26 | nvo 155 | color 0 | alpha 0 | ssim 0 | psnr 27 | wps 1.6 | wpb 1 | bsz 1 | num_updates 5000 | best_loss 2
> 2020-10-14 18:12:47 | INFO | fairseq.checkpoint_utils | saved checkpoint /home/tuotuoli/work/output/NSVF/Bike/checkpoint1.pt (epoch 1 @ 5000 updates, score 2.0) (writing took 0.06901660500079743 seconds)
> 2020-10-14 18:12:47 | INFO | fairnr.modules.encoder | pruning...
> 2020-10-14 18:12:47 | INFO | fairnr.modules.encoder | pruning done. # of voxels before: 155, after: 148 voxels
> 2020-10-14 18:12:47 | INFO | fairnr.models.nsvf | half the global voxel size 0.1600 -> 0.0800
> 2020-10-14 18:12:47 | INFO | fairnr.modules.encoder | splitting...
> Building EasyOctree done. total #nodes = 367, terminal #nodes = 148 (time taken 0.041932 s)
> 2020-10-14 18:12:47 | INFO | fairnr.modules.encoder | splitting done. # of voxels before: 148, after: 1184 voxels
> 2020-10-14 18:12:47 | INFO | fairnr_cli.train | reset the trainer at 5000
> 2020-10-14 18:12:47 | INFO | fairseq.trainer | NOTE: your device may support faster training with --fp16
> 2020-10-14 18:12:47 | INFO | fairnr.models.nsvf | reduce the raymarching step size 0.0200 -> 0.0100
> Building EasyOctree done. total #nodes = 2479, terminal #nodes = 1184 (time taken 0.345659 s)
> 2020-10-14 18:12:48 | INFO | train_inner | epoch 002: 1 / 5000 loss=37, vox=0, stp=0, tvo=0, asf=42, ash=42, nvo=1184, color=0, alpha=0, wps=0.1, ups=0.07, wpb=2, bsz=1, num_updates=5001, lr=0.00096666, gnorm=106, train_wall=1, wall=1848

For now, if I turned of "--use-octree", the wired error disappeared. The strange thing is this argument works for many scenes (which I have tested before). You can turn-off octree for now and use the simple ray voxel intersection.

**Comment by TuotuoLi** [comment]
Thanks for the update and confirming the issue :)

**Comment by TuotuoLi** [comment]
I think I have a quick fix here.
I tried this originally at another line,but did not fixed the problem, after you told me to disable octree, I find there is an other line need fix.
```

def build_easy_octree(points, half_voxel):
    from fairnr.clib._ext import build_octree
    coords = ((points _+ **torch.min(points,0).values.unsqueeze(0) + 0.00001_**)/ half_voxel).floor_().long()     # works easier in int space
    residual = (points - coords.float() * half_voxel).mean(0, keepdim=True)
    ranges = coords.max(0)[0] - coords.min(0)[0]
    depths = torch.log2(ranges.max().float()).ceil_().long() - 1
    center = (coords.max(0)[0] + coords.min(0)[0])//2
    centers, children = build_octree(center, coords, int(depths))
    centers = centers.float() * half_voxel + residual   # transform back to float
    return centers, children


```

**Comment by MultiPath** [comment]
> I think I have a quick fix here.
> I tried this originally at another line,but did not fixed the problem, after you told me to disable octree, I find there is an other line need fix.
> 
> ```
> def build_easy_octree(points, half_voxel):
>     from fairnr.clib._ext import build_octree
>     coords = ((points - torch.min(points,0).values.unsqueeze(0) + 0.00001)/ half_voxel).floor_().long()     # works easier in int space
>     residual = (points - coords.float() * half_voxel).mean(0, keepdim=True)
>     ranges = coords.max(0)[0] - coords.min(0)[0]
>     depths = torch.log2(ranges.max().float()).ceil_().long() - 1
>     center = (coords.max(0)[0] + coords.min(0)[0])//2
>     centers, children = build_octree(center, coords, int(depths))
>     centers = centers.float() * half_voxel + residual   # transform back to float
>     return centers, children
> ```
> 
> The other place is line 241
> ` old_coords = ((point_xyz - torch.min(point_xyz,0).values.unsqueeze(0) + 0.00001)/ quarter_voxel).floor_().long()`

Thanks for looking into this! 
Yeah, it seems my original discretization hits some edge cases and failed to work. Will commit changes based on your suggestions.

**Comment by TuotuoLi** [comment]
Thanks for sharing the code for us to learn!!

**Comment by MultiPath** [comment]
> Thanks for sharing the code for us to learn!!

Hi, I have merged the code with similar changes. Thanks
